experiment/measurer/test_experiment_coverage_utils.py

```diff
@@ -13,32 +13,301 @@
+                        TIMESTAMP))
+
+    fuzzer_ids = df_container.function_df['fuzzer'].unique()
+    benchmark_ids = df_container.function_df['benchmark'].unique()
+    function_ids = df_container.function_df['function'].unique()
+
+    assert len(df_container.segment_df) == NUM_COVERED_SEGMENTS_IN_COV_SUMMARY
+    assert len(df_container.function_df) == NUM_FUNCTION_IN_COV_SUMMARY
+
+    for fuzzer_id in fuzzer_ids:
+        # Check if type recorded for the ID id is "fuzzer".
+        assert (df_container.name_df[df_container.name_df['id'] == fuzzer_id]
+                ['type'].item() == "fuzzer")
+        # Check if fuzzer ids match in "function" data frame and "name" data
+        # frame.
+        assert (fuzzer_id == df_container.name_df[df_container.name_df['name']
+                                                  == FUZZER]['id'].item())
+    for benchmark_id in benchmark_ids:
+        # Check if type recorded for the ID id is "benchmark".
+        assert (df_container.name_df[df_container.name_df['id'] == benchmark_id]
+                ['type'].item() == "benchmark")
+        # Check if benchmark ids match in "function" data frame and "name" data
+        # frame.
+        assert (benchmark_id == df_container.name_df[
+            df_container.name_df['name'] == BENCHMARK]['id'].item())
+
+    for function_id in function_ids:
+        # check all function names in the data frames are already known.
+        assert (set(
+            df_container.name_df[df_container.name_df['id'] == function_id]
+            ['name'].unique()).issubset(FUNCTION_NAMES))
+        # Check if function ids match in "function" data frame and "name" data
+        # frame.
+        assert (df_container.name_df[df_container.name_df['id'] == function_id]
+                ['type'].item() == "function")
+
+
+def test_data_frame_container_remove_redundant_duplicates(fs):
+    """Tests that remove_redundant_duplicates removes all duplicates entries in
+    name_df and segment_df."""
+
+    summary_json_file = get_test_data_path(SUMMARY_JSON_FILE)
+    fs.add_real_file(summary_json_file, read_only=False)
+
+    df_container = (experiment_coverage_utils.
+                    extract_segments_and_functions_from_summary_json(
+                        summary_json_file, BENCHMARK, FUZZER, TRIAL_ID,
+                        TIMESTAMP))
+
+    # Drop duplicates using remove_redundant_duplicates().
+    df_container.remove_redundant_duplicates()
+
+    # length of data frame after calling remove_redundant_duplicates().
+    df_length = len(df_container.segment_df)
+
+    # Force dropping of duplicates again.
+    df_after_drop = df_container.segment_df.drop_duplicates(
+        subset=df_container.segment_df.columns.difference(['time']))
+
+    # Data frame length after for drop.
+    df_length_after_drop = len(df_after_drop)
+
+    # assert length didn't change.
+    assert (df_length - df_length_after_drop) == 0
```

Comment:
Is there a better way to check whether there are duplicates, e.g., using the pandas function `duplicates()`?

Reply:
There is function called `duplicated()` in pandas, which return a list as long as the data frame with `TRUE` or `FALSE` values indicating wether another similar row exists or not. It looks a bit messy if I use that for this test (hard to understand).